pkg/scale/decode.go

```diff
@@ -478,51 +478,77 @@ func (ds *decodeState) decodeBool(dstv reflect.Value) (err error) {
 
 // decodeUint will decode unsigned integer
 func (ds *decodeState) decodeUint(dstv reflect.Value) (err error) {
-	b, err := ds.ReadByte()
+	const maxUint32 = ^uint32(0)
+	const maxUint64 = ^uint64(0)
+	prefix, err := ds.ReadByte()
 	if err != nil {
-		return
+		return fmt.Errorf("reading from buffer: %w", err)
 	}
 
 	in := dstv.Interface()
 	temp := reflect.New(reflect.TypeOf(in))
 	// check mode of encoding, stored at 2 least significant bits
-	mode := b & 3
-	switch {
-	case mode <= 2:
-		var val int64
-		val, err = ds.decodeSmallInt(b, mode)
+	mode := prefix % 4
+	var value uint
+	switch mode {
+	case 0:
+		value = uint(prefix >> 2)
+	case 1:
+		var buf byte
+		buf, err = ds.ReadByte()
 		if err != nil {
-			return
+			return fmt.Errorf("reading from buffer: %w", err)
 		}
-		temp.Elem().Set(reflect.ValueOf(val).Convert(reflect.TypeOf(in)))
-		dstv.Set(temp.Elem())
-	default:
-		// >4 byte mode
-		topSixBits := b >> 2
-		byteLen := uint(topSixBits) + 4
-
+		value = uint(binary.LittleEndian.Uint16([]byte{prefix, buf}) >> 2)
+		if value <= 0b0011_1111 || value > 0b0111_1111_1111_1111 {
+			return ErrU64OutOfRange
```

Comment:
nit add context with it

```suggestion
			return fmt.Errorf("%w: %d (%16b)", ErrU64OutOfRange, value, value)
```

Note I'm not 100% sure `%16b` works but it might just work